python/nav/Snmp/pynetsnmp.py

```diff
@@ -85,30 +88,86 @@ class Snmp(object):
     """
 
     def __init__(
-        self, host, community="public", version="1", port=161, retries=3, timeout=1
+        self,
+        host: str,
+        community: str = "public",
+        version: Union[str, int] = "1",
+        port: Union[str, int] = 161,
+        retries: int = 3,
+        timeout: int = 1,
+        # SNMPv3-only options
+        sec_level: Optional[SecurityLevel] = None,
+        auth_protocol: Optional[AuthenticationProtocol] = None,
+        sec_name: Optional[str] = None,
+        auth_password: Optional[str] = None,
+        priv_protocol: Optional[PrivacyProtocol] = None,
+        priv_password: Optional[str] = None,
     ):
         """Makes a new Snmp-object.
 
         :param host: hostname or IP address
         :param community: community (password), defaults to "public"
         :param port: udp port number, defaults to "161"
+        :param sec_level: SNMPv3 security level
+        :param auth_protocol: SNMPv3 authentication protocol
+        :param sec_name: SNMPv3 securityName
+        :param auth_password: SNMPv3 authentication password
+        :param priv_protocol: SNMPv3 privacy protocol
+        :param priv_password: SNMPv3 privacy password
 
         """
-
+        self.handle = None
         self.host = host
         self.community = str(community)
         self.version = str(version)
         if self.version == '2':
             self.version = '2c'
-        if self.version not in ('1', '2c'):
+        if self.version not in ('1', '2c', '3'):
             raise UnsupportedSnmpVersionError(self.version)
         self.port = int(port)
         self.retries = retries
         self.timeout = timeout
 
+        self.sec_level = SecurityLevel(sec_level) if sec_level else None
+        self.auth_protocol = (
+            AuthenticationProtocol(auth_protocol) if auth_protocol else None
+        )
+        self.sec_name = sec_name
+        self.auth_password = auth_password
+        self.priv_protocol = PrivacyProtocol(priv_protocol) if priv_protocol else None
+        self.priv_password = priv_password
+        if self.version == "3":
+            self._verify_snmpv3_params()
+
         self.handle = _MySnmpSession(self._build_cmdline())
         self.handle.open()
 
+    def _verify_snmpv3_params(self):
```

Comment:
As @stveit suggested in the automerged one, a dataclass for the SNMPv3 config sounds better and better. Then this method could be on that class.

The instance would be stored as a whole in the `Snmp` init, which would make for more verbose code but probably better since theere is only one Snmp-class.

Reply:
I think this is a good idea(tm).  But there is now a bunch of PRs based on this that are waiting to be merged, and all of them need to change as a result of this.

May I suggest doing a refactor PR to fix all the affected bits once they've all been merged? I can add it as a task to #1177 .

Reply:
Good idea.